jax/_src/extend/ffi.py

```diff
@@ -140,3 +148,110 @@ def _ir_attribute(obj: Any) -> ir.Attribute:
+  will be called directly on batched inputs (where the batch axes are the
+  leading dimensions). Additionally, the callbacks should return outputs that
+  have corresponding leading batch axes. If ``vectorized`` is ``False`` (the
+  default behavior), transforming this ``ffi_call`` under :func:`~jax.vmap` will
+  result in a :func:`~jax.lax.scan` with the ``ffi_call`` in the body.
+
+  Args:
+    ffi_target_name: the name of the XLA FFI custom call target that was
+      registered using :func:`~jaxlib.xla_client.register_custom_call_target`.
+    result_shape_dtypes: an object, or sequence of objects, with ``shape`` and
+      ``dtype`` attributes which are expected to match the shape and dtype of
+      the custom call output or outputs. :class:`~jax.ShapeDtypeStruct` is often
+      used to define the elements of ``result_shape_dtypes``.
+    *args: the arguments passed to the custom call.
+    vectorized: boolean specifying whether the callback function can operate in
+      a vectorized manner, as described above.
+    **kwargs: keyword arguments that are passed as named attributes to the
+      custom call using XLA's FFI interface.
+
+  Returns:
+    One or more :class:`~jax.Array` objects whose shapes and dtypes match
+    ``result_shape_dtypes``.
+  """
+  if isinstance(result_shape_dtypes, Sequence):
+    multiple_results = True
+    result_types = result_shape_dtypes
+  else:
+    multiple_results = False
+    result_types = (result_shape_dtypes,)
+  map(_check_shape_dtype, result_types)
+  result_avals = tuple(core.ShapedArray(x.shape, x.dtype) for x in result_types)
+  results = ffi_call_p.bind(
+      *args,
+      result_avals=result_avals,
+      vectorized=vectorized,
+      ffi_target_name=ffi_target_name,
+      **kwargs,
+  )
+  if multiple_results:
+    return results
+  else:
+    return results[0]
+
+
+def ffi_call_abstract_eval(
+    *avals_in,
+    result_avals: tuple[core.ShapedArray, ...],
+    ffi_target_name: str,
+    vectorized: bool,
+    **kwargs: Any,
+):
+  del avals_in, ffi_target_name, vectorized, kwargs
+  return result_avals
+
+
+def ffi_call_jvp(*args, ffi_target_name, **kwargs):
+  del args, kwargs
+  raise ValueError(
+      f"The FFI call to `{ffi_target_name}` cannot be differentiated. "
+      "You can use `jax.custom_jvp` or `jax.custom_jvp` to add support.")
+
+
+def ffi_call_transpose(*args, ffi_target_name, **kwargs):
+  del args, kwargs
+  raise ValueError(
+      f"The FFI call to `{ffi_target_name}` cannot be differentiated. "
+      "You can use `jax.custom_jvp` or `jax.custom_jvp` to add support.")
+
+
+def ffi_call_lowering(
+    ctx: mlir.LoweringRuleContext,
+    *operands: ir.Value,
+    result_avals: tuple[core.ShapedArray, ...],
+    ffi_target_name: str,
+    vectorized: bool,
+    **kwargs: Any,
+) -> Sequence[ir.Value]:
+  del result_avals, vectorized
+  return ffi_lowering(ffi_target_name)(ctx, *operands, **kwargs)
+
+
+ffi_call_p = core.Primitive("ffi_call")
```

Comment:
Love the potential for reducing boilerplate!
How does this work when `ffi_call` is called multiple times over the life of a process? There seems to be only one `ffi_call_p` declared globally, does that mean `ffi_call` will overwrite the abstract eval rule, impl, etc.? Or are you planning to have a global `ffi_call_p_registry` or similar that adds a new primitive every time `ffi_call` is called? Or is that not necessary somehow?

Reply:
Good question! There's actually only ever need for one primitive because all of the relevant parameters are passed at bind time. You'll see that all the rules take the target name as a parameter, for example. This means that the same primitive can be reused for any number of calls to the same (or many different) FFI targets.

This is actually the same pattern which is used for the callback primitives ([here's the primitive for `pure_callback`, for example](https://github.com/google/jax/blob/d577e29998c8b9a3db5d835bbf9af8c13f453a01/jax/_src/callback.py#L43)), and all other higher order primitives (`custom_vjp`, etc.).

Reply:
I see thanks, so if I'm understanding this correctly, `ffi_call_p` assumes a new "persona" every time `ffi_call` is invoked and that works fine.

Reply:
Exactly! If you inspect the jaxpr with an `ffi_call` in it, you'll see something like:

```
ffi_call[
  target_name=my_ffi_call_target_name
  result_avals=(ShapedArray(float32[5,3]), ...)
  vectorized=True
  a_parameter_that_i_set=1.5
]
```